src/sched/lthread.c

```diff
@@ -567,8 +567,13 @@ int _lthread_resume(struct lthread* lt)
     if (lt->yield_cb)
     {
         lt->yield_cb(lt->yield_cbarg);
+        // Yielding while holding a ticket lock may cause deadlock
+        SGXLKL_ASSERT(lt->ticketlocks_held == 0);
     }
 
+    sched->current_lthread = NULL;
```

Comment:
did you intend to move these two lines?

Reply:
Yes, without that any code run in the callback is not able to get the current thread. That makes it impossible for the unlock to decrement the counter.

I did not send this as a PR because I’m uncertain if that has any other negative side effects. I don’t believe it should be possible for the lthread strict to be deallocated before here, but I’m not completely sure. There may be a race condition in the cloned thread deallocation if there are multiple ethreads. That said, the resume function expects the lthread object to be still live because it accesses a field, so this shouldn’t make things worse...

Reply:
Bad things definitely happen with this here. See:

https://dev.azure.com/sgx-lkl/sgx-lkl/_build/results?buildId=1549&view=logs&j=4dfb742b-8e24-5ebe-1c3b-99535daead16

for some segfaults in addition to the assertions being triggered.